### PR TITLE
feat: keystone ownership transfer [KS-558]

### PR DESCRIPTION
**Pull request by ecPablo**
[KS-558](https://smartcontract-it.atlassian.net/browse/KS-558)

* Creates 2 new changesets `AcceptAllOwnershipsProposal` and `TransferAllOwnership` to accept and transfer ownership of the 4 keystone contracts: `CapabilitiesRegistry, KeystoneForwarder, OCR3Capability, FeedsConsumer `. I'm not 100% familiar with the keystone product so please let me know if there are contracts missing here.
* Adds some utils to fetch typed contracts from the address book.
* Reuses the common changeset from mcms to accept and transfer ownership to generate the ownership acceptance proposal.
* Renamed some parameter names like `TimelocksPerChain` to `OwnersPerChain`. This is to make the point that is not always necessary to have the timelock as an owner. There can be products that might not use the timelock contract

[KS-558]: https://smartcontract-it.atlassian.net/browse/KS-558?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: deployment/common/changeset/transfer_ownership.go [hunk 1]
```diff
@@ -15,18 +16,18 @@ type OwnershipTransferrer interface {
 }
 
 type TransferOwnershipConfig struct {
-	// TimelocksPerChain is a mapping from chain selector to the timelock contract address on that chain.
-	TimelocksPerChain map[uint64]common.Address
+	// OwnersPerChain is a mapping from chain selector to the timelock contract address on that chain.
```
Comment (graham-chainlink) [inline deployment/common/changeset/transfer_ownership.go]:
```suggestion
	// OwnersPerChain is a mapping from chain selector to the owner's contract address on that chain.
```

File: deployment/keystone/types.go [hunk 2]
```diff
@@ -20,10 +20,12 @@ import (
...
+	KeystoneForwarder         deployment.ContractType = "KeystoneForwarder"         // https://github.com/smartcontractkit/chainlink/blob/50c1b3dbf31bd145b312739b08967600a5c67f30/contracts/src/v0.8/keystone/KeystoneForwarder.sol#L90
+	OCR3Capability            deployment.ContractType = "OCR3Capability"            // https://github.com/smartcontractkit/chainlink/blob/50c1b3dbf31bd145b312739b08967600a5c67f30/contracts/src/v0.8/keystone/OCR3Capability.sol#L12
+	FeedConsumer              deployment.ContractType = "FeedConsumer"              // no type and a version in contract https://github.com/smartcontractkit/chainlink/blob/89183a8a5d22b1aeca0ade3b76d16aa84067aa57/contracts/src/v0.8/keystone/KeystoneFeedsConsumer.sol#L1
+	RBACTimelock              deployment.ContractType = "RBACTimelock"              // no type and a version in contract https://github.com/smartcontractkit/ccip-owner-contracts/blob/main/src/RBACTimelock.sol
+	ProposerManyChainMultiSig deployment.ContractType = "ProposerManyChainMultiSig" // no type and a version in contract https://github.com/smartcontractkit/ccip-owner-contracts/blob/main/src/RBACTimelock.sol
```
Comment (gustavogama-cll) [inline deployment/keystone/types.go]: [nit] shouldn't the comment refer to `https://github.com/smartcontractkit/ccip-owner-contracts/blob/main/src/ManyChainMultiSig.sol`?

Comment (ecPablo) [reply deployment/keystone/types.go]: Yes good catch ! https://github.com/smartcontractkit/chainlink/pull/15451/commits/40901342680d7f387c886fd4ef234b2d07e38adf

**Review by graham-chainlink — DISMISSED** [review]

**Review by RodrigoAD — DISMISSED** [review]

**Review by graham-chainlink — DISMISSED** [review]

**Review by akhilchainani — DISMISSED** [review]

File: deployment/keystone/contract_set.go [hunk 3]
```diff
@@ -92,3 +92,19 @@ func DeployForwarder(lggr logger.Logger, chain deployment.Chain, ab deployment.A
...
+	consumerDeploy := KeystoneFeedsConsumerDeployer{lggr: lggr}
+	consumerResp, err := consumerDeploy.deploy(DeployRequest{Chain: chain})
+	if err != nil {
+		return fmt.Errorf("failed to deploy FeedsConsumer: %w", err)
+	}
+	err = ab.Save(chain.Selector, consumerResp.Address.String(), consumerResp.Tv)
+	if err != nil {
+		return fmt.Errorf("failed to save FeedsConsumer: %w", err)
+	}
+	lggr.Infof("Deployed %s chain selector %d addr %s", consumerResp.Tv.String(), chain.Selector, consumerResp.Address.String())
```
Comment (jmank88) [inline deployment/keystone/contract_set.go]: This log line is pretty deep in the call stack. Typically things are cleaner when we return errors from plain `funcs` and log only from method `funcs`, which have a contextual logger attached and which we expect to have these sorts of side effects. How would things look if we logged this line from the caller, and dropped the `lggr logger.Logger` argument?

File: deployment/keystone/changeset/accept_ownership.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	mcmsProposers, err := proposersFromAddrBook(addrBook, chain)
+	if err != nil {
+		return deployment.ChangesetOutput{}, err
+	}
+
+	// Initialize the Contracts slice
+	var ownershipAcceptors []changeset.OwnershipAcceptor
+
+	// Append CapabilitiesRegistry contracts
+	for _, capReg := range capRegs {
```
Comment (krehermann) [inline deployment/keystone/changeset/accept_ownership.go]: all of these for loops can be simplified to variadic ~ 
```
ownership = append(ownership, foo...)
ownership = append(ownership, bar...)
...
```

File: deployment/keystone/changeset/accept_ownership.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	// Construct the configuration
+	cfg := changeset.AcceptOwnershipConfig{
+		OwnersPerChain: map[uint64]common.Address{
+			// Assuming there is only one timelock per chain.
+			chainSelector: timelocks[0].Address(),
+		},
+		ProposerMCMSes: map[uint64]*ccipowner.ManyChainMultiSig{
+			// Assuming there is only one MCMS proposer per chain.
```
Comment (krehermann) [inline deployment/keystone/changeset/accept_ownership.go:78]: what does it mean to have only 1 proposer? is there some background material i can read?

File: deployment/keystone/changeset/accept_ownership.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/ethereum/go-ethereum/common"
+
+	ccipowner "github.com/smartcontractkit/ccip-owner-contracts/pkg/gethwrappers"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	"github.com/smartcontractkit/chainlink/deployment/common/changeset"
+)
+
+// AcceptAllOwnershipsProposal creates a MCMS proposal to call accept ownership on all the Keystone contracts in the address book.
+func AcceptAllOwnershipsProposal(e deployment.Environment, chainSelector uint64, minDelay time.Duration) (deployment.ChangesetOutput, error) {
```
Comment (krehermann) [inline deployment/keystone/changeset/accept_ownership.go]: change this signature to match the generic changeset func and add a type assertion [eg](https://github.com/smartcontractkit/chainlink/blob/86d5f9993ef208ef0a310a781b198563f4ae7d55/deployment/keystone/changeset/append_node_capbilities.go#L11)

File: deployment/keystone/changeset/deploy_consumer.go [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import (
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	kslib "github.com/smartcontractkit/chainlink/deployment/keystone"
+)
+
+// DeployFeedsConsumer deploys the FeedsConsumer contract to the chain with the given chainSelector.
+func DeployFeedsConsumer(env deployment.Environment, chainSelector uint64) (deployment.ChangesetOutput, error) {
```
Comment (krehermann) [inline deployment/keystone/changeset/deploy_consumer.go]: add type assertion as changeset func as above

File: deployment/keystone/changeset/transfer_ownership.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	"github.com/smartcontractkit/chainlink/deployment/common/changeset"
+)
+
+// TransferAllOwnership transfers ownership of all Keystone contracts in the address book to the existing timelock.
+func TransferAllOwnership(e deployment.Environment, chainSelector uint64) (deployment.ChangesetOutput, error) {
```
Comment (krehermann) [inline deployment/keystone/changeset/transfer_ownership.go]: type assertion here, too

File: deployment/keystone/changeset/transfer_ownership.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	consumers, err := feedsConsumersFromAddrBook(addrBook, chain)
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed to fetch feeds consumers: %w", err)
+	}
+
+	// Initialize the Contracts slice
+	var ownershipTransferrers []changeset.OwnershipTransferrer
+
+	// Append CapabilitiesRegistry contracts
+	for _, capReg := range capRegs {
```
Comment (krehermann) [inline deployment/keystone/changeset/transfer_ownership.go]: can be variadic here, too

**Review by krehermann — APPROVED** [review]

**Review by akhilchainani — APPROVED** [review]

**Pull request merged**